### ISSUE TITLE
Replace `crypto-js` library with Deno `crypto` APIs

### ISSUE THREAD

**Issue by jcs224** [jcs224/oak_sessions#29]

**Comment by c0per** [comment]
After a bit of research, I find this issue more complicated than I thought.

By default (currently), `CryptoJS` uses AES CBC mode, which can be done using `crypto` api. This mode of AES needs a key and an iv (initialization vector) to work.

`CryptoJS` will treat the string "key" passed in as a "passphrase". Then use "EVP Key Derivation Function" to get the real "key" and "iv" needed to encrypt using AES.

However, web `crypto` api only takes in "key" and "iv". It won't do this derivation for you. See the api: https://developer.mozilla.org/en-US/docs/Web/API/AesCbcParams

Of course we can write our own derivation function (It's [`EVP_BytesToKey` function in OpenSSL](https://www.openssl.org/docs/man3.0/man3/EVP_BytesToKey.html)). It's only a few lines of code but it uses md5 hash. But there's no way to compute md5 in `crypto` api because md5 is considered unsafe and deprecated. 😅Another dependency to do md5 needs to be added to replace `CryptoJS`.

**Comment by c0per** [comment]
There's lots of articles about moving from `crypto-js` to `crypto` api like [this one](https://qwtel.com/posts/software/replacing-cryptojs-with-web-cryptography/).

I'm not sure if adding another (probably smaller) dependency and some dangerously looking derivation code to remove `crypto-js` is worth it.

Or we can use [project like this](https://github.com/decent-bet/crypto-evpkdf) to do the derivation and encrypt/decrypt using web `crypto` api.

**Comment by jcs224** [comment]
We might not need to rely on web crypto alone, after some digging it turns out Deno has its own crypto interface in the standard library that augments what web crypto has. If a hashing algorithm isn't in standard web crypto, it uses a Rust/Wasm substitute.

https://github.com/denoland/deno_std/tree/main/crypto

**Comment by c0per** [comment]
> We might not need to rely on web crypto alone, after some digging it turns out Deno has its own crypto interface in the standard library that augments what web crypto has. If a hashing algorithm isn't in standard web crypto, it uses a Rust/Wasm substitute.
> 
> https://github.com/denoland/deno_std/tree/main/crypto

That looks promising. I'll write some code to try it out. Thanks.

**Comment by jcs224** [comment]
I did a bit of research on this myself, these are some string encrypt/decrypt functions I made. Now the key has to be 16, 24, or 32 characters long (for 128, 192, or 256-bit AES-CBC), and it prepends a random IV based on `crypto.getRandomValues()`. Thinking this should be secure enough? All encryption is web crypto, just using base64-encoded string (like the current way) to store the encrypted bytes.

```ts
const te = (s : string) => new TextEncoder().encode(s)
const td = (d: Uint8Array) => new TextDecoder().decode(d)
import { encode as be, decode as bd } from "https://deno.land/std@0.149.0/encoding/base64.ts";

async function encryptString(key_string : string, data_string : string) {
  const key_encoded = te(key_string)
  const key = await crypto.subtle.importKey('raw', key_encoded.buffer, 'AES-CBC', true, ['encrypt', 'decrypt'])

  const iv = crypto.getRandomValues(new Uint8Array(16))
  const iv_base64_string = be(iv)
  const iv_base64_shortened = iv_base64_string.substring(0, iv_base64_string.length - 2)

  const encrypted = await crypto.subtle.encrypt({ name: 'AES-CBC', iv}, key, te(data_string))
  const encrypted_bytes = new Uint8Array(encrypted)
  const encrypted_data_base64_string = be(encrypted_bytes)

  return iv_base64_shortened + encrypted_data_base64_string
}

async function decryptString(key_string : string, data_string_payload : string) {
  const key_encoded = te(key_string)
  const key = await crypto.subtle.importKey('raw', key_encoded.buffer, 'AES-CBC', true, ['encrypt', 'decrypt'])

  const data_string_encrypted = data_string_payload.substring(22)
  const iv_base64_string = data_string_payload.substring(0, 22)
  const iv = bd(iv_base64_string)

  const decrypted = await crypto.subtle.decrypt({ name: 'AES-CBC', iv}, key, bd(data_string_encrypted))
  const decrypted_bytes = new Uint8Array(decrypted)
  const decrypted_text = td(decrypted_bytes)

  return decrypted_text
}
```

**Comment by c0per** [comment]
> I did a bit of research on this myself, these are some string encrypt/decrypt functions I made. Now the key has to be 16, 24, or 32 characters long (for 128, 192, or 256-bit AES-CBC), and it prepends a random IV based on `crypto.getRandomValues()`. Thinking this should be secure enough? All encryption is web crypto, just using base64-encoded string (like the current way) to store the encrypted bytes.
> 
> ```ts
> const te = (s : string) => new TextEncoder().encode(s)
> const td = (d: Uint8Array) => new TextDecoder().decode(d)
> import { encode as be, decode as bd } from "https://deno.land/std@0.149.0/encoding/base64.ts";
> 
> async function encryptString(key_string : string, data_string : string) {
>   const key_encoded = te(key_string)
>   const key = await crypto.subtle.importKey('raw', key_encoded.buffer, 'AES-CBC', true, ['encrypt', 'decrypt'])
> 
>   const iv = crypto.getRandomValues(new Uint8Array(16))
>   const iv_base64_string = be(iv)
>   const iv_base64_shortened = iv_base64_string.substring(0, iv_base64_string.length - 2)
> 
>   const encrypted = await crypto.subtle.encrypt({ name: 'AES-CBC', iv}, key, te(data_string))
>   const encrypted_bytes = new Uint8Array(encrypted)
>   const encrypted_data_base64_string = be(encrypted_bytes)
> 
>   return iv_base64_shortened + encrypted_data_base64_string
> }
> 
> async function decryptString(key_string : string, data_string_payload : string) {
>   const key_encoded = te(key_string)
>   const key = await crypto.subtle.importKey('raw', key_encoded.buffer, 'AES-CBC', true, ['encrypt', 'decrypt'])
> 
>   const data_string_encrypted = data_string_payload.substring(22)
>   const iv_base64_string = data_string_payload.substring(0, 22)
>   const iv = bd(iv_base64_string)
> 
>   const decrypted = await crypto.subtle.decrypt({ name: 'AES-CBC', iv}, key, bd(data_string_encrypted))
>   const decrypted_bytes = new Uint8Array(decrypted)
>   const decrypted_text = td(decrypted_bytes)
> 
>   return decrypted_text
> }
> ```

I made a PR #31, using the `crypto` and `base64` module in deno std to achive the "`CryptoJS` style" AES encoding.

The biggest difference between this code and the PR is that the PR one uses the EVP Key Derivation Function (`evpkdf`) to obtain key and iv from the string passphrase passed in by user just like before.

Another difference is the PR code will put "__Salted", salt, cipherText together and return a base64 string just like the `CryptoJS` do. And it will also try to parse the incoming string to get the salt in `decrypt` function. You can checkout the code to be clear about this.

**Comment by c0per** [comment]
Actually, I saw som people warn that `CryptoJS` be default uses `evpkdf` with only 1 iteration. So the key and iv actully comes from md5 one time the passphrase, and that's not safe. (I'm not sure if it's true of how unsafe it is. I'm not an expert.)


Maybe we should switch to another safer encryption, and that will be some kind of breaking. All the existing cookie session will be invalid due to switch in decryption algorithm.

**Comment by jcs224** [comment]
Nice work!

In the name of security, I think we should harden our encryption as much as possible as long as it doesn't significantly hinder performance on the server. Even if that means a breaking change. Since we're storing a string visible to the user, it should be tamper-proof at all costs. From my reading, `md5` and `SHA-1` aren't quite up to the task. `SHA-384` looked good in terms of not having the collidable or length-extendable vulnerabilities, and it's faster than the other ones on 64-bit machines (64-bit is a requirement to run Deno anyway).

On this vein, another breaking change is, I want to make a key mandatory for CookieStore (#26). Not encrypting the cookie is dangerous, not just for the user browsing the site, but all other users on a site using CookieStore.

**Comment by jcs224** [comment]
Put a couple changes in a branch.

https://github.com/c0per/oak_sessions/compare/crypto-api...jcs224:oak_sessions:sha-384

1. Switched to SHA-384, this will break existing cookies but it works without any other changes!
2. In case the decryption fails (which it will for existing cookies), create a new session

I just had an idea. If we don't want to destroy existing data, we could migrate it by when they visit the site, decrypting their MD5-based keys, re-encrypting with SHA-384, then saving that result to the cookie.

**Comment by c0per** [comment]
> Put a couple changes in a branch.
> 
> [c0per/oak_sessions@crypto-api...jcs224:oak_sessions:sha-384](https://github.com/c0per/oak_sessions/compare/crypto-api...jcs224:oak_sessions:sha-384)
> 
> 1. Switched to SHA-384, this will break existing cookies but it works without any other changes!
> 2. In case the decryption fails (which it will for existing cookies), create a new session
> 
> I just had an idea. If we don't want to destroy existing data, we could migrate it by when they visit the site, decrypting their MD5-based keys, re-encrypting with SHA-384, then saving that result to the cookie.

That seems great. I think of a way to do this "fallback" decryption.
This seems to be a noob question... But how can I commit on top of you latest commit? It looks like it's your pr built on top of mine, which I can't access. May be you should merge first or something? So I can commit the code for md5 fallback on that latest pr.

Thanks.

**Comment by c0per** [comment]
Maybe you can set me as one of the "maintainers"? Github allow maintainers to edit existing PR if the creator of the PR set that way.

**Comment by jcs224** [comment]
I went ahead and merged everything, feel free to make another PR.

**Comment by jcs224** [comment]
Fixed by #31 ,#32, #33
